frontend/appflowy_flutter/lib/workspace/presentation/home/menu/sidebar/workspace/_sidebar_workspace_item_list.dart

```diff
@@ -0,0 +1,209 @@
+
+class WorkspacesMenu extends StatelessWidget {
+  const WorkspacesMenu({
+    super.key,
+    required this.userProfile,
+    required this.currentWorkspace,
+    required this.workspaces,
+  });
+
+  final UserProfilePB userProfile;
+  final UserWorkspacePB currentWorkspace;
+  final List<UserWorkspacePB> workspaces;
+
+  @override
+  Widget build(BuildContext context) {
+    return BlocListener<UserWorkspaceBloc, UserWorkspaceState>(
+      listener: (context, state) {
+        final result = state.createWorkspaceResult;
+        if (result != null) {
+          final message = result.fold(
+            (s) => LocaleKeys.workspace_createSuccess.tr(),
+            (e) => '${LocaleKeys.workspace_createFailed.tr()}: $e',
+          );
+          showMessageToast(message);
+        }
+      },
+      child: Column(
+        crossAxisAlignment: CrossAxisAlignment.start,
+        mainAxisSize: MainAxisSize.min,
+        children: [
+          // user email
+          Padding(
+            padding: const EdgeInsets.symmetric(vertical: 8, horizontal: 4),
+            child: Row(
+              children: [
+                FlowyText.medium(
+                  _getUserInfo(),
+                  fontSize: 12.0,
+                  overflow: TextOverflow.ellipsis,
+                  color: Theme.of(context).hintColor,
+                ),
+                const Spacer(),
+                FlowyButton(
+                  useIntrinsicWidth: true,
+                  text: const FlowySvg(FlowySvgs.add_m),
+                  onTap: () {
+                    _showCreateWorkspaceDialog(context);
+                    PopoverContainer.of(context).closeAll();
+                  },
+                ),
+              ],
+            ),
+          ),
+          for (final workspace in workspaces) ...[
+            _WorkspaceMenuItem(
+              workspace: workspace,
+              userProfile: userProfile,
+              isSelected: workspace.workspaceId == currentWorkspace.workspaceId,
+            ),
+            const VSpace(4.0),
+          ],
+        ],
+      ),
+    );
+  }
+
+  String _getUserInfo() {
+    if (userProfile.email.isNotEmpty) {
+      return userProfile.email;
+    }
+
+    if (userProfile.name.isNotEmpty) {
+      return userProfile.name;
+    }
+
+    return LocaleKeys.defaultUsername.tr();
+  }
+
+  Future<void> _showCreateWorkspaceDialog(BuildContext context) async {
+    if (context.mounted) {
+      await NavigatorTextFieldDialog(
+        title: LocaleKeys.workspace_create.tr(),
+        value: '',
+        hintText: '',
+        autoSelectAllText: true,
+        onConfirm: (name, context) async {
+          final request = CreateWorkspacePB.create()..name = name;
+          final result = await UserEventCreateWorkspace(request).send();
+          final message = result.fold(
+            (s) => LocaleKeys.workspace_createSuccess.tr(),
+            (e) => '${LocaleKeys.workspace_createFailed.tr()}: ${e.msg}',
+          );
```

Comment:
Application logic belongs in a BLoC

Reply:
"The WorkspacesMenu widget will be disposed of before this logic runs. I originally used bloc, but it caused a Provider issue, so I switched to this approach.

<img width="1241" alt="Screenshot 2024-03-03 at 23 28 42" src="https://github.com/AppFlowy-IO/AppFlowy/assets/11863087/4501cd87-72d3-421d-9186-d3e742adfb0e">